Refactor Tileset Traversal

austinEng opened the pull request:
- Traversal and functions relevant to traversal have been moved into a new file `Cesium3DTilesetTraversal.js`
- Tileset has an added property `baseScreenSpaceError` which defines the screen space error which must be met before LOD skipping starts
- Traversal has been generalized as a pipeline of tree traversals. The base traversal is a DFS, the leaves of that are used to initialize a BFS for the skipping traversal which internally uses a DFS. The traversal has been refactored so each traversal method has functions `execute`, `visit`, `getChildren`, `shouldVisit`, and `leafHandler`.
- To prevent tiles from being processed or requested twice (which would happen when the leaf of one traversal becomes the root of the following, `_requestedFrame ` and `_lastVisitedFrame` have been added to  `Cesium3DTile`
- Some loops after the old traversal have been removed and simplified by keeping track of `_ancestorWithContent ` and `_ancestorWithLoadedContent ` during the top-down traversal. It is no longer necessary to do a bottom-up traversal.
- `_sse` has been renamed to `_screenSpaceError`
- Setting`skipLODs = false` now just sets `baseScreenSpaceError` to `_maximumScreenSpaceError` and runs a base traversal.
- The refactored traversal better manages the `_hasMixedContent` flag so that the stencil test is only applied when necessary. Previously, it would always set `_hasMixedContent` to `true` if we had to go up the tree to select a tile. It is more correct to do this only when `_selectionDepth > 0` which implies a selected tile has a selected ancestor. Tests have been updated to account for this.

lilleyse commented on Source/Scene/Cesium3DTilesetTraversal.js:
```diff
@@ -0,0 +1,741 @@
+            var length = leaves.length;
+            for (var i = 0; i < length; ++i) {
+                tileset._desiredTiles.push(leaves.get(i));
+            }
```
When `skipLODs` is false, should the `bivariateVisibilityTest` ever be on? I am seeing some cases where it is.

lilleyse commented:
After testing the old performance/ behavior is still there which is good, minus the low-lod geometry blips that happened sometimes. I like the way the traversal code is organized now.

lilleyse commented:
The new code doesn't use the child union optimization. Is that intentional?

austinEng commented on Source/Scene/Cesium3DTilesetTraversal.js:
It's needed if `tile.hasTilesetContent === true`

austinEng commented:
@lilleyse Updated based on comments.

> When would stack.trim be undefined?
Whoops. Just an old thing

> The new code doesn't use the child union optimization. Is that intentional?
Accidentally removed. Added back.

> tile._screenSpaceError > baseScreenSpaceError will always be true by this point.
Fixed. The one with maximum screen space error isn't always true. We need the check if `tile.hasTilesetContent === true`

pjcozzi commented:
@austinEng have you ran a few JavaScript profiles to make sure this didn't introduce any high CPU or memory usage?

pjcozzi commented:
There's a few CI test failures: https://travis-ci.org/AnalyticalGraphicsInc/cesium/builds/228357786

austinEng commented:
@lilleyse Updated. 

Spent a lot of time trying to solve what seemed to be bugs in the traversal. This turned out to be problems with translucency.  Translucent commands were being put in the `CESIUM_3D_TILE` render pass and taken out of the translucency pass. Tilesets with translucent geometry seemed to be randomly occluding content behind them. @bagnell I've added a check so that translucent commands do not get put in the `CESIUM_3D_TILE` pass.

@pjcozzi Performance and memory usage is about the same.

lilleyse commented on Source/Scene/Cesium3DTilesetTraversal.js:
I don't completely see why the children are updated here.

lilleyse commented on Source/Scene/Cesium3DTilesetTraversal.js:
Should be defined in the class.

lilleyse commented on Source/Scene/Cesium3DTilesetTraversal.js:
This last visited check is pretty much the same throughout, why not put back in `visitTile`?

lilleyse commented on Source/Scene/Cesium3DTilesetTraversal.js:
Also this goes to the second point, are `visitStart` and `visitEnd` needed? If `InternalBaseTraversal.prototype.visitStart` is changed based on other comments, `visitStart` would be the same throughout. The code in `visitEnd` can be incorporated in `visit` along with my comment right above.

lilleyse commented on Source/Scene/Cesium3DTilesetTraversal.js:
Should there be a `touchedFrame`? Some things will be touched twice - first in `getChildren` and then `visitTile`.

lilleyse commented on Source/Scene/Cesium3DTilesetTraversal.js:
Since `shouldVisit()` will return false for any tile with renderable content, this code would not got reached for those tiles. This would create a situation `allLoaded` is true even if one of its descendants with renderable content is not ready.

austinEng commented on Source/Scene/Cesium3DTilesetTraversal.js:
1. `visitStart`
2. `getChildren`
3. if no children, `leafHandler`
4. `visitEnd`

sometimes the leaf handler checks `tile._lastVisitedFrame !== this.frameState.frameNumber` so we don't add tiles twice

austinEng commented on Source/Scene/Cesium3DTilesetTraversal.js:
`leafHandler` checks if at least one child is visible for the children union bound optimization

lilleyse commented on Source/Scene/Cesium3DTilesetTraversal.js:
Ok, leave a comment about that.

austinEng commented on Source/Scene/Cesium3DTilesetTraversal.js:
I think you're right, but I think we should then add to the condition below

`tile._screenSpaceError <= baseScreenSpaceError && !tile.hasTilesetContent`

so that the base traversal doesn't stop at the root of external tilesets

lilleyse commented on Source/Scene/Cesium3DTilesetTraversal.js:
For that case it can stay, but do you think the others can be simplified?

lilleyse commented on Source/Scene/Cesium3DTilesetTraversal.js:
Never mind, I see now. Maybe a short `isVisited` helper would be cleaner.

austinEng commented on Source/Scene/Cesium3DTilesetTraversal.js:
Is it worth adding this? The other flags are important because they prevent tiles from getting put onto a list twice, but this just moves the tile in the doubly linked list.

lilleyse commented on Source/Scene/Cesium3DTilesetTraversal.js:
Well it's a bit like the sse check.

austinEng commented:
@lilleyse Updated

lilleyse commented:
Thanks @austinEng, the recent updates look good.
